experimental/ComplexReflectionGroups/src/complex_reflection_group.jl

```diff
@@ -0,0 +1,182 @@
+        elseif Cmodel == :CHEVIE
+            matgrp = complex_reflection_group_CHEVIE(t)
+        else
+            error("Specified model not found")
+        end
+
+        # set attributes that are already known from type
+        set_attribute!(matgrp, :order, order(C))
+        set_attribute!(matgrp, :is_complex_reflection_group, true)
+        set_attribute!(matgrp, :complex_reflection_group_type, C)
+        set_attribute!(matgrp, :complex_reflection_group_model, [Cmodel])
+        set_attribute!(matgrp, :is_irreducible, true)
+
+        # add to list
+        push!(component_groups, matgrp)
+        push!(modellist, model)
+    end
+
+    if length(component_groups) == 1
+        # treat this as a special case because direct_product([G]) will have type direct
+        # product which looks weird for a single group.
+        return component_groups[1]
+    else
+        matgrp = direct_product(component_groups)
+
+        set_attribute!(matgrp, :order, order(G))
+        set_attribute!(matgrp, :is_complex_reflection_group, true)
+        set_attribute!(matgrp, :complex_reflection_group_type, G)
+        set_attribute!(matgrp, :complex_reflection_group_model, modellist)
+        set_attribute!(matgrp, :is_irreducible, false)
+
+        return matgrp
+    end
+
+end
+
+# Convenience constructors
+complex_reflection_group(i::Int, model::Symbol=:default) = complex_reflection_group(ComplexReflectionGroupType(i), model)
+
+complex_reflection_group(m::Int, p::Int, n::Int, model::Symbol=:default) = complex_reflection_group(ComplexReflectionGroupType(m,p,n), model)
+
+complex_reflection_group(X::Vector, model::Symbol=:default) = complex_reflection_group(ComplexReflectionGroupType(X), model)
+
+
+
+
+function complex_reflection_group_type(G::MatrixGroup)
+    if has_attribute(G, :complex_reflection_group_type)
+        return get_attribute(G, :complex_reflection_group_type)
+    end
+    return nothing
+    # this should be upgraded later to work with a general matrix group (identifying the
+    # type from scratch is not so easy though)
+end
+
+function complex_reflection_group_model(G::MatrixGroup)
+    if has_attribute(G, :complex_reflection_group_model)
+        return get_attribute(G, :complex_reflection_group_model)
+    end
+    return nothing
+end
+
+function complex_reflection_group_dual(W::MatrixGroup)
+
+    WD = matrix_group([transpose(matrix(w^-1)) for w in gens(W)])
+
+    set_attribute!(WD, :order, order(W))
+    set_attribute!(WD, :is_complex_reflection_group, true)
+    set_attribute!(WD, :complex_reflection_group_type, complex_reflection_group_type(W))
+
+    return WD
+
+end
+
+
+###########################################################################################
+# Checking if a matrix group is a complex reflection group.
+###########################################################################################
+function is_complex_reflection_group(G::MatrixGroup{T}) where T <: QQAlgFieldElem
```

Comment:
This functions only checks whether the chosen generators of the group are reflections, but not whether the group is a reflection group. Or am I missing something?

Reply:
Yes, indeed, and this will be remarked in the documentation. For the general thing you'd have to check through all subsets of reflection whether they generate and this can be a lot of work. Also, for computer algebra, it should be the correct notion that the fixed generators are reflections. As I said, this will be remarked in the documentation and if someone wants to implement the general thing it will be another function.

Reply:
Hm, I am not sure this is very OSCAR like, where we try to stay as close to the mathematics without too much surprising behaviour for a non-expert. Being a complex reflection group is a mathematical concept independent of the chosen generators. The natural name for checking this of an arbitrary matrix group would be `is_complex_reflection_group`, but this would be already taken by this function here.

A similar situation would be ideals, which for OSCAR always come with a fixed family/set of generators. But `is_principal(I)` for such an ideal `I` will check whether it can be generated by one element and not if `ngens(I) == 1` holds.

Maybe a suggestion would be to rename this into something internal, maybe put an `_` in front of it?

Reply:
In similar situations we discussed using `is_created_as_*` for this exact use-case. But we postponed it and made the function internal with a leading underscore for the time being

Reply:
Why not name it generated by reflections? And reserve the other one for the mathematical test?

Reply:
Alright alright alright alright alright alright, you'll get your honest reflection group. It's more consistent than what I did but I would never admit this here. Also, it's actually not so bad to check as I thought: just take the set of all reflections and see if the subgroup generated by this is equal to the group. Only finding all reflections could be optimized (as commented in my code) by using conjugacy classes. But I'll leave that to whoever is interested. :-)

Reply:
Done in [d611586](https://github.com/oscar-system/Oscar.jl/pull/3342/commits/d6115864ea13756ba06124f31d8e875eceefb059). I will expand the ComplexReflection structure a bit to save the parent in case of a group element (or just GL if it's a matrix) so that this avoids the G(matrix(w)), which may involve computations which are actually not necessary.